### ISSUE TITLE
If in a plist a payload in `PayloadContent` does not have a `PayloadDescription`, updating `jamfpro_macos_configuration_profile_plist.description` deletes the `PayloadContent` array & breaks the profile object on Jamf instance

### ISSUE THREAD

**Issue by w0de** [deploymenttheory/terraform-provider-jamfpro#351]
1. Define, plan/apply a new profile with one payload in the `PayloadContent` array. The payload must not have `PayloadDescription` set. Do set a `jamfpro_macos_configuration_profile_plist.description` value.
2. Change the `jamfpro_macos_configuration_profile_plist.description` value. Plan/apply.
3. After apply, run another plan. The plan will show that current state has an empty `PayloadContent` array, and will plan to recreate it. Also newly empty: `Payload{DisplayName,Identifier,Scope,Type,Identifier,UUID,Version,Enabled}` (Version and Enabled are set to 0 and false, rather than empty).
4. Attempting to apply this plan - or even just attempting to view the profile in Jamf's GUI - results in an error in Jamf's server log. It can't parse the profile object. At this point the profile must be destroyed and recreated.

```
Caused by: org.springframework.orm.jpa.JpaSystemException: java.lang.IllegalArgumentException: The given string value: {"forced": {"plist": "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<!DOCTYPE plist PUBLIC \"-//Apple//DTD PLIST 1.0//EN\" \"http://www.apple.com/DTDs/PropertyList-1.0.dtd\">\n<plist version=\"1.0\">\n  <dict>\n    <key>CloudManagementEnrollmentToken</key>\n    <string>a3b7a8fb-0b3a-4e40-a632-e1a58770e884</string>\n  </dict>\n</plist>\n"}, "payloadType": "com.apple.ManagedClient.preferences", "payloadUUID": "54119F64-6504-460B-9DAC-069B01F43C29", "unknownKeys": {"com.apple.ManagedClient.preferences": {"PayloadScope": {"type": "string"}, "PayloadEnabled": {"type": "false", "value": false}} }, "payloadVersion": 1, "preferenceDomain": "com.google.Chrome", "payloadIdentifier": "54119F64-6504-460B-9DAC-069B01F43C29", "payloadDisplayName": "Custom Settings", "payloadOrganization": "JAMF Software"} cannot be transformed to Json object; nested exception is org.hibernate.HibernateException: java.lang.IllegalArgumentException: The given string value: {"forced": {"plist": "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<!DOCTYPE plist PUBLIC \"-//Apple//DTD PLIST 1.0//EN\" \"http://www.apple.com/DTDs/PropertyList-1.0.dtd\">\n<plist version=\"1.0\">\n  <dict>\n    <key>CloudManagementEnrollmentToken</key>\n    <string>a3b7a8fb-0b3a-4e40-a632-e1a58770e884</string>\n  </dict>\n</plist>\n"}, "payloadType": "com.apple.ManagedClient.preferences", "payloadUUID": "54119F64-6504-460B-9DAC-069B01F43C29", "unknownKeys": {"com.apple.ManagedClient.preferences": {"PayloadScope": {"type": "string"}, "PayloadEnabled": {"type": "false", "value": false}} }, "payloadVersion": 1, "preferenceDomain": "com.google.Chrome", "payloadIdentifier": "54119F64-6504-460B-9DAC-069B01F43C29", "payloadDisplayName": "Custom Settings", "payloadOrganization": "JAMF Software"} cannot be transformed to Json object
```

This probably has something to do with `plist.NormalizePayloadState` or other common/plist func(s). Digging in further - can also provide more logs if requested.

**Comment by w0de** [comment]
I can only reproduce this (minus one time which I'm starting to doubt happened 🤪) with profiles containing a `PayloadContent` dict nested `PayloadContent`. For example:

```xml
<dict>
		<key>PayloadContent</key>
		<array>
			<dict>
				<key>PayloadContent</key>
				<dict>
					<key>com.cloudflare.1dot1dot1dot1.macos</key>
					<dict>
						<key>Forced</key>
						<array>
							<dict>
								<key>mcx_preference_settings</key>
								<dict>
									<key>SUAutomaticallyUpdate</key>
									<false/>
									<key>SUEnableAutomaticChecks</key>
									<false/>
									<key>SUFeedURL</key>
									<string>https://127.0.0.1</string>
								</dict>
							</dict>
						</array>
					</dict>
				</dict>
				<key>PayloadEnabled</key>
				<true/>
				<key>PayloadIdentifier</key>
				<string>186e9655-affb-412d-ad80-8e0991d76d97</string>
				<key>PayloadType</key>
				<string>com.apple.ManagedClient.preferences</string>
				<key>PayloadUUID</key>
				<string>186e9655-affb-412d-ad80-8e0991d76d97</string>
				<key>PayloadVersion</key>
				<integer>1</integer>
			</dict>
		</array>
```

Also, the top-level `PayloadDescription` doesn't seem to matter - can be omitted, empty, or have a value: adding a `resource.description` value to existing profile still triggers the bug.

**Comment by w0de** [comment]
This appears to be a Jamf bug triggered by our the described, and very particular, input (nested PayloadContent _dict_, no initial description set). 

Because:
- I've confirmed the payloads string sent to the server in request that patches `description` isn't missing its PayloadContent array and isn't otherwise malformed.
- Damningly, the bug occurs 5-10m _after_ Terraform's patch. Immediately after TF's patch request, both GUI and API show the correct payload and the correct, updated `description` attribute. Then - one surmises - an automated server-side Jamf job processes the profile, and deletes PayloadContent.

I think the best workaround is to make description a computed property. 

Jamf will set it from the plist's top-level PayloadDescription, if there is one. If there isn't, it will set a generic "Settings for <PayloadType>..." description (though not always - at least only when there is but one Payload in PayloadContent, and maybe also not when the PayloadType isn't known/implemented in their GUI abstraction layer). 

This will also fix a more minor bug I've discovered while investigating: apparently one can't delete `description`, only set/update it. Will also set `name` computed: changing it causes Jamf - probs the same automated job - to rewrite the remote plist's top-level `PayloadDisplayName`.

Opening a PR for this. Bonus: it will also fix

**Comment by smithjw** [comment]
@w0de One thing I'm hoping you've fixed in this PR, is differences between the source file having a new line and the output not having it. 

Currently bypassing the following error by passing the mobileconfig into the `payloads` with a `chomp()` function

`Error: in 'jamfpro_macos_configuration_profile.Nudge - Workplace Team (testing)': plist line count mismatch: source has 81 lines, formatted has 80 lines, check for trailing lines and whitespace`

**Comment by ShocOne** [comment]
hi there, there are a number of fixes implemented for macos config profiles in the v0.1.8 release that should handle nested dictionaries correctly when performing diff suppression. i've also added some handling for trailing whitespace as well. [https://github.com/deploymenttheory/terraform-provider-jamfpro/blob/main/internal/resources/common/configurationprofiles/plist/diffsuppression.go](url)

i can successfully deploy this

`<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
	<dict>
		<key>PayloadContent</key>
		<array>
			<dict>
				<key>PayloadContent</key>
				<dict>
					<key>com.cloudflare.1dot1dot1dot1.macos</key>
					<dict>
						<key>Forced</key>
						<array>
							<dict>
								<key>mcx_preference_settings</key>
								<dict>
									<key>SUAutomaticallyUpdate</key>
									<false/>
									<key>SUEnableAutomaticChecks</key>
									<false/>
									<key>SUFeedURL</key>
									<string>https://127.0.0.1</string>
								</dict>
							</dict>
						</array>
					</dict>
				</dict>
				<key>PayloadEnabled</key>
				<true/>
				<key>PayloadIdentifier</key>
				<string>186e9655-affb-412d-ad80-8e0991d76d97</string>
				<key>PayloadType</key>
				<string>com.apple.ManagedClient.preferences</string>
				<key>PayloadUUID</key>
				<string>186e9655-affb-412d-ad80-8e0991d76d97</string>
				<key>PayloadVersion</key>
				<integer>1</integer>
			</dict>
		</array>
		<key>PayloadDescription</key>
		<string/>
		<key>PayloadDisplayName</key>
		<string>github-issue-test</string>
		<key>PayloadEnabled</key>
		<true/>
		<key>PayloadIdentifier</key>
		<string>913B8754-7F50-4342-8143-8FCE7843BEFB</string>
		<key>PayloadOrganization</key>
		<string>Deployment Theory</string>
		<key>PayloadRemovalDisallowed</key>
		<true/>
		<key>PayloadScope</key>
		<string>System</string>
		<key>PayloadType</key>
		<string>Configuration</string>
		<key>PayloadUUID</key>
		<string>913B8754-7F50-4342-8143-8FCE7843BEFB</string>
		<key>PayloadVersion</key>
		<integer>1</integer>
	</dict>
</plist>`
deployment looks like this
![image](https://github.com/user-attachments/assets/9a934eb6-43b5-4c08-ac9a-cecbecc3b91e)
a re-run of terraform gives this 

![image](https://github.com/user-attachments/assets/67e936ff-8fb3-4591-82eb-f767f957d6c2)

I'm going to close to issue but feel free to raise new issues with other config profiles if you find unexpected behaviour.

**Comment by smithjw** [comment]
@ShocOne i think that is someone else's profile but the usecase is the same!

Does this also apply to the profile generator resource?
And if so, are you able to provide a HCL example showing how to correctly nest the dicts to produce the same output as your screenshot?
